crates/core/c8y_api/src/json_c8y.rs

```diff
@@ -236,95 +238,167 @@ impl SourceInfo {
         Self { id, source_type }
     }
 }
+
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
 #[serde(rename_all = "camelCase")]
 pub struct C8yCreateAlarm {
+    /// Alarm type, default is "ThinEdgeAlarm".
+    #[serde(rename = "type")]
+    pub alarm_type: String,
+
+    /// None for main device, Some for child device.
     #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(rename = "externalSource")]
     pub source: Option<SourceInfo>,
 
-    pub severity: String,
+    /// If None, the message is meant to clear the alarm,
+    /// Some means to create a new alarm.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(flatten)]
+    pub data: Option<C8yCreateAlarmData>,
```

Comment:
Good point. It can be a flag `clear: true` for example. I'll consider.

Reply:
@albinsuresh at first, I was aligned with you, asking @rina23q to group all related data into a single `struct`.

But I have to acknowledge that there is a nice property of this two-parts message. This gives a nice way to represent either an alarm or a message to clear the alarm. There is a similar issue with commands, where one needs to be cautious before [de-serializing a command message](https://github.com/thin-edge/thin-edge.io/blob/main/crates/extensions/c8y_mapper_ext/src/converter.rs#L740) because the command structs are enable to represent the clearing case.

Reply:
I think @albinsuresh and @didier-wenzek are on the same page here. But I believe that @albinsuresh says `C8yCreateAlarm` doesn't need to have a sub-struct as it's used only for **serializing** a struct to c8y's way.

In c8y's API, [`"status": "CLEARED"`](https://cumulocity.com/guides/10.4.6/reference/alarms/#alarm-applicationvndcomnsncumulocityalarmjson) is the key. We don't need to use it now as we clear an alarm by SmartREST, however, once we move to JSON over MQTT, we will need this field. 

So I'll change the current struct to align more with c8y.

Reply:
@didier-wenzek I was also fully aligned with you on that decomposition of the `ThinEdgeAlarm` struct, precisely for that reason that you highlighted. That's why even the former struct had this decomposition using a sub-struct representing the payload. But, here, as @rina23q also highlighted, my concern was about replicating the same style for the `C8yCreateAlarm` struct, which looked unnecessary, as the representation of a "clear alarm message" was not an empty sub-struct in C8y.

IMO, instead of overloading the `C8yCreateAlarm` struct, by making `severity` and `text` fields optional, to handle clearing as well, it'd be better have an entirely different struct, a `C8yClearAlarm`, to represent that case.

Reply:
> as the representation of a "clear alarm message" was not an empty sub-struct in C8y.

Correct. I start thinking rather making two different struct for creating an alarm and clearing an alarm is the right way. Why?
1. As @albinsuresh said, "clearing an alarm" is not an empty sub-struct. It has ["status": "CLEARED"](https://cumulocity.com/guides/10.4.6/reference/alarms/#alarm-applicationvndcomnsncumulocityalarmjson) or not.
2. But as I said, having **optional** `severity`, `text` and `time` are also wrong as they are mandatory to create an alarm.  

So I start feeling we are talking about two different (but similar in terms of alarm) data models and trying to keep it in one struct. I would go to have two structs if others agree with me. Otherwise, I will not change.

Reply:
Discussed with @albinsuresh, and this model satisfies both 1) and 2) as I mentioned in the above comment. I'll change.

```rust
enum C8yAlarm {
    Create(C8yCreateAlarm),
    Clear(C8yClearAlarm)
    
}

struct C8yCreateAlarm {
    // all necessary fields, and all are mandatory.
    
}

struct C8yClearAlarm {
    alarm_type: &str,
    source: Option<SourceInfo>,
}
```

```rust
let c8y_alarm = C8yAlarm::from(&tedge_alarm);
```